src/libraries/System.Private.CoreLib/src/System/Reflection/NullabilityInfoContext.cs

```diff
@@ -499,39 +469,126 @@ private void CheckGenericParameters(NullabilityInfo nullability, MemberInfo meta
 
                     for (int i = 0; i < genericArguments.Length; i++)
                     {
-                        if (genericArguments[i].IsGenericParameter)
-                        {
-                            int index = i + 1;
-                            NullabilityInfo n = GetNullabilityInfo(metaMember, genericArguments[i], genericArguments[i].GetCustomAttributesData(), ref index);
-                            nullability.GenericTypeArguments[i].ReadState = n.ReadState;
-                            nullability.GenericTypeArguments[i].WriteState = n.WriteState;
-                        }
-                        else
-                        {
-                            UpdateGenericArrayElements(nullability.GenericTypeArguments[i].ElementType, metaMember, genericArguments[i]);
-                        }
+                        CheckGenericParameters(nullability.GenericTypeArguments[i], metaMember, genericArguments[i], reflectedType);
                     }
                 }
-                else
+                else if (nullability.ElementType is { } elementNullability && metaType.IsArray)
                 {
-                    UpdateGenericArrayElements(nullability.ElementType, metaMember, metaType);
+                    CheckGenericParameters(elementNullability, metaMember, metaType.GetElementType()!, reflectedType);
                 }
             }
         }
 
-        private void UpdateGenericArrayElements(NullabilityInfo? elementState, MemberInfo metaMember, Type metaType)
+        private void UpdateGenericParameterNullability(NullabilityInfo nullability, Type genericParameter, Type? reflectedType)
         {
-            if (metaType.IsArray && elementState != null
-                && metaType.GetElementType()!.IsGenericParameter)
+            Debug.Assert(genericParameter.IsGenericParameter);
+
+            if (reflectedType is not null
+                && !genericParameter.IsGenericMethodParameter
+                && TryUpdateGenericTypeParameterNullabilityFromReflectedType(nullability, genericParameter, reflectedType, reflectedType))
+            {
+                return;
+            }
+
+            var state = NullabilityState.Unknown;
+            if (CreateParser(genericParameter.GetCustomAttributesData()).ParseNullableState(0, ref state))
             {
-                Type elementType = metaType.GetElementType()!;
-                int index = 0;
-                NullabilityInfo n = GetNullabilityInfo(metaMember, elementType, elementType.GetCustomAttributesData(), ref index);
-                elementState.ReadState = n.ReadState;
-                elementState.WriteState = n.WriteState;
+                nullability.ReadState = state;
+                nullability.WriteState = state;
+                return;
             }
         }
 
+        private bool TryUpdateGenericTypeParameterNullabilityFromReflectedType(NullabilityInfo nullability, Type genericParameter, Type context, Type reflectedType)
```

Comment:
This is the most complicated new logic. We're trying to figure out nullability information for a generic parameter based on a type that derives from the class that declared that parameter. When deriving from a generic type, the compiler annotates the derived type with nullable metadata about the parameters used, for example:

```C#
  [Nullable(new byte[] {(byte) 0, (byte) 1, (byte) 1})]
  internal class NonNullableValuesDictionary : Dictionary<string, object>
  {
  }

  [Nullable(new byte[] {(byte) 0, (byte) 1, (byte) 2})]
  internal class NullableValuesDictionary : Dictionary<string, object>
  {
  }
```

Sometimes we need to track through multiple parameters, e. g. in a case like this:
```C#
class TaggedList<TTag, TElement> : List<TElement> { }

class TaggedListOfStrings<TTag> : TaggedList<TTag, string> { }
```